### PR TITLE
ci: prevent validating transitive edr & hardhat versions

### PR DESCRIPTION
**Pull request by Wodann**
The HH3 regression benchmark now runs to completion but fails on an overzealous validity check that is triggered by transitive EDR or Hardhat dependencies not being updated.

This PR fixes that by only requiring direct Hardhat package inclusions (and their EDR version) to be updated to the local version.

- **Should be included**: `1inch-swap-vm` -> `hardhat ${bumpedVersion}` -> `edr ${version}-local.{sha}`
- **Previously false positive; now no longer checked**: `1inch-swap-vm` -> third-party-dep -> `hardhat ${oldVersion}` -> `edr. ${oldVersion}`

This is a temporary CI check that's meant to help us catch failures in the benchmark setup. Once we are confident that the benchmark is running with correct versions, we'll remove the step from the workflow.

See failing CI job: https://github.com/NomicFoundation/edr/actions/runs/27950141616/job/82769063058

### CODE REVIEW

File: .github/workflows/hh3-regression-benchmark.yml [hunk 1]
```diff
@@ -222,45 +222,74 @@ jobs:
...
+                  if (parent === d) throw new Error("package.json for " + name + " not found above " + main);
+                  d = parent;
+                }
+              }
+              try {
+                const hhPkg = pkgJsonPath("hardhat", dir);
+                const edrPkg = pkgJsonPath("@nomicfoundation/edr", path.dirname(hhPkg));
+                const ver = (p) => JSON.parse(fs.readFileSync(p, "utf8")).version;
+                process.stdout.write(ver(hhPkg) + " " + ver(edrPkg));
+              } catch (e) { process.stderr.write(String(e.message || e)); }
```
[the remark this answers is not in the document]
Comment (Wodann) [reply .github/workflows/hh3-regression-benchmark.yml:262]: The platform package is always the same version as the EDR version

File: .github/workflows/hh3-regression-benchmark.yml [hunk 2]
```diff
@@ -222,45 +222,74 @@ jobs:
...
+            ' "$d" 2>/tmp/edr-resolve-err)
 
-            # yarn Plug'n'Play scenarios have no node_modules; fall back to the lockfile.
-            if [ -f "$d/.pnp.cjs" ] || [ -f "$d/.pnp.loader.mjs" ]; then
-              if grep -rqs -- "$EDR_VER" "$d/yarn.lock"; then
+            if [ -z "${edr_ver:-}" ]; then
```
[the remark this answers is not in the document]
Comment (Wodann) [reply .github/workflows/hh3-regression-benchmark.yml:265]: The platform package is always the same version as the EDR version

File: .github/workflows/hh3-regression-benchmark.yml [hunk 3]
```diff
@@ -222,45 +222,74 @@ jobs:
...
+              fail=1
+              continue
+            fi
+
+            found=1
+            if [ "$edr_ver" != "$EDR_VER" ]; then
+              echo "::error::$id: hardhat@$hh_ver resolves @nomicfoundation/edr@$edr_ver, expected $EDR_VER"
+              fail=1
+            else
+              echo "OK: $id -> hardhat@$hh_ver, @nomicfoundation/edr@$edr_ver"
```
[the remark this answers is not in the document]
Comment (Wodann) [reply .github/workflows/hh3-regression-benchmark.yml:283]: The platform package is always the same version as the EDR version

File: .github/workflows/hh3-regression-benchmark.yml [hunk 4]
```diff
@@ -222,45 +222,74 @@ jobs:
...
-                 -o -path '*/node_modules/@nomicfoundation/edr-linux-x64-gnu/package.json' \) \
-              2>/dev/null)
+            id=$(basename "$d")
+
+            # Resolve the EDR that THIS scenario's hardhat loads. Works across
+            # npm/pnpm/yarn node_modules layouts (Node realpaths pnpm symlinks).
+            # Nested EDR copies pulled by other dependents (e.g. @1inch/solidity-utils,
+            # hardhat-network-helpers) are intentionally ignored — only hardhat's
+            # EDR (the build under test) matters for the benchmark.
+            read -r hh_ver edr_ver < <(node -e '
```
[the remark this answers is not in the document]
Comment (Wodann) [reply .github/workflows/hh3-regression-benchmark.yml:237]: The platform package is always the same version as the EDR version

**Comment by Wodann** [comment]
This PR doesn't alter the actual code, so the JS Solidity test runner benchmark regression is a false positive: https://github.com/NomicFoundation/edr/actions/runs/27996153350/job/82858882487?pr=1502

File: .github/workflows/hh3-regression-benchmark.yml [hunk 5]
```diff
@@ -222,45 +222,74 @@ jobs:
...
+                  d = parent;
+                }
+              }
+              try {
+                const hhPkg = pkgJsonPath("hardhat", dir);
+                const edrPkg = pkgJsonPath("@nomicfoundation/edr", path.dirname(hhPkg));
+                const ver = (p) => JSON.parse(fs.readFileSync(p, "utf8")).version;
+                process.stdout.write(ver(hhPkg) + " " + ver(edrPkg));
+              } catch (e) { process.stderr.write(String(e.message || e)); }
+            ' "$d" 2>/tmp/edr-resolve-err)
```
Comment (anaPerezGhiglia) [inline .github/workflows/hh3-regression-benchmark.yml]:
```suggestion
            ' "$d" 2>/tmp/edr-resolve-err)  || true
```

`read` returns a non-zero exit status when it reaches EOF without seeing a newline delimiter. The `node -e` script writes its output with `process.stdout.write(...)` (line 261), which emits no trailing `\n`, so read exits non-zero even when it successfully assigns `hh_ver`/`edr_ver`. Under the step's `set -euo pipefail` (line 207), that aborts the whole step on the first scenario in the loop — Check B never actually validates anything.

Adding a newline to the node output isn't enough: the genuine resolution-failure path writes only to stderr (line 262), leaving stdout empty, so read still hits EOF immediately and returns non-zero.

By addig the `|| true` we are neutralizing read's exit status; the existing `[ -z "${edr_ver:-}" ]` check then handles the empty case.

**Review by anaPerezGhiglia — COMMENTED** [review]
I understand this validation step is meant to be temporary, so I don't want to over-engineer it. That said, I found it genuinely hard to review and validate as-is.

The dense bash + embedded Node, the multiple resolution fallbacks, and the subtle `set -euo pipefail` interactions (like the read/EOF issue I flagged inline) made the control flow hard to verify by inspection, I had to lean on Claude to trace the edge cases.

Even as a temporary check, moving the Node resolution logic into a standalone `.github/scripts/...` file we can run and test locally (instead of inlining it in YAML) would make this much easier to review and maintain. 

This is not blocking if we want to land and iterate, but I wanted to flag that the current form is hard to reason about and easy to break.

**Comment by Wodann** [comment]
> I understand this validation step is meant to be temporary, so I don't want to over-engineer it. That said, I found it genuinely hard to review and validate as-is.

I agree it's hard to reason about. As this is only temporary code intended to validate that the workflow (which only runs from `main`) is working; I chose not to move the JS to the `scripts/` directory (like I did for permanent JS code), nor add tests.

It felt like over-engineering something that's really only meant to catch errors in the benchmark initialization; and will be removed as soon as everything is working.

Normally I'd not submit this code, but workflows only work once they're merged to `main`, which makes iterating this cumbersome.

**Review by anaPerezGhiglia — APPROVED** [review]
